### PR TITLE
Add color to request and response body when content-type is json

### PR DESCRIPTION
**Pull request by Amirhy**
## :camera: Screenshots
Screen Shot 1401-09-07 at 14 21 45
Screen Shot 1401-09-07 at 14 22 09

## :page_facing_up: Context

I have been using chucker for a long time. But the gray texts in the body of the requests and responses bothered me and my teammates's eyes and also made it difficult for us to read the texts properly.
So I decided to color the text of body of requests and responses.
Although I've only done it for application/json content-type and I'll support the rest of them soon.
I also fixed a bug that if there is no HeaderItem type inside the adapter items and you search for a text in the body of requests and responses, the text does not highlight properly.

## :pencil: Changes

- Implement colorful text for body of request and response when content-type is application/json

## :hammer_and_wrench: How to test
I just write some tests for json formatting in SpanUtilTest class that you can run them and also you can do functional testing.
## :stopwatch: Next steps
Implement this feature for other content-types , for example application/xml

### CODE REVIEW

**Comment by cortinico** [comment]
Thanks for sending this @Amirhy. I won't be able to review this till sometime next week as I'm a bit busy at the moment 👍

**Comment by Amirhy** [comment]
Hi @cortinico. Hope you 're doing well.
Did you have free time to review my PR?

**Comment by cortinico** [comment]
> Did you have free time to review my PR?

Sorry I was quite busy this week. I'll have to look into this over the next weekend.
In the meanwhile @ArjanSM do you want to do a pass here?

**Comment by ArjanSM** [comment]
> > Did you have free time to review my PR?
> 
> Sorry I was quite busy this week. I'll have to look into this over the next weekend. In the meanwhile @ArjanSM do you want to do a pass here?

@cortinico I've been a little occupied today. I'll take a look at it tomorrow.

**Review by ArjanSM — COMMENTED** [review]
Thanks for the PR @Amirhy 🏅 
I've made a few suggestions.
Also, I personally feel that it would be nice if you could create different PRs (one for text coloring feature and the other for the Text highlighting bug) or the reference to the feature issue and the bug in the PR description.

**Comment by Amirhy** [comment]
> Thanks for the PR @Amirhy 🏅 I've made a few suggestions. Also, I personally feel that it would be nice if you could create different PRs (one for text coloring feature and the other for the Text highlighting bug) or the reference to the feature issue and the bug in the PR description.

Thanks for your time @ArjanSM. I will fix them as soon as possible.

**Comment by Amirhy** [comment]
> Thanks for the PR @Amirhy 🏅 I've made a few suggestions. Also, I personally feel that it would be nice if you could create different PRs (one for text coloring feature and the other for the Text highlighting bug) or the reference to the feature issue and the bug in the PR description.

Ok I removed my changes that fix the mentioned bug, and I will create another PR for them.

**Comment by Amirhy** [comment]
@ArjanSM  kindly reminder

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadAdapter.kt [hunk 1]
```diff
@@ -36,19 +38,25 @@ internal class TransactionBodyAdapter : RecyclerView.Adapter<TransactionPayloadV
         holder.bind(items[position])
     }
 
-    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TransactionPayloadViewHolder {
+    override fun onCreateViewHolder(
+        parent: ViewGroup,
+        viewType: Int
+    ): TransactionPayloadViewHolder {
```
Comment (ArjanSM) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadAdapter.kt:44]: nit: maybe revert this change.
Though I understand what you were trying here.

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadAdapter.kt [hunk 2]
```diff
@@ -36,19 +38,25 @@ internal class TransactionBodyAdapter : RecyclerView.Adapter<TransactionPayloadV
...
         val inflater = LayoutInflater.from(parent.context)
         return when (viewType) {
             TYPE_HEADERS -> {
-                val headersItemBinding = ChuckerTransactionItemHeadersBinding.inflate(inflater, parent, false)
+                val headersItemBinding =
+                    ChuckerTransactionItemHeadersBinding.inflate(inflater, parent, false)
```
Comment (ArjanSM) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadAdapter.kt:49]: nit: this too. unless this was a detekt suggestion.

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadAdapter.kt [hunk 3]
```diff
@@ -36,19 +38,25 @@ internal class TransactionBodyAdapter : RecyclerView.Adapter<TransactionPayloadV
...
                 TransactionPayloadViewHolder.HeaderViewHolder(headersItemBinding)
             }
             TYPE_BODY_LINE -> {
-                val bodyItemBinding = ChuckerTransactionItemBodyLineBinding.inflate(inflater, parent, false)
+                val bodyItemBinding =
+                    ChuckerTransactionItemBodyLineBinding.inflate(inflater, parent, false)
```
Comment (ArjanSM) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadAdapter.kt:54]: nit: maybe revert this change.

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadAdapter.kt [hunk 4]
```diff
@@ -36,19 +38,25 @@ internal class TransactionBodyAdapter : RecyclerView.Adapter<TransactionPayloadV
...
                 TransactionPayloadViewHolder.BodyLineViewHolder(bodyItemBinding)
             }
             else -> {
-                val imageItemBinding = ChuckerTransactionItemImageBinding.inflate(inflater, parent, false)
+                val imageItemBinding =
+                    ChuckerTransactionItemImageBinding.inflate(inflater, parent, false)
```
Comment (ArjanSM) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadAdapter.kt:59]: nit: revert this change.

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadAdapter.kt [hunk 5]
```diff
@@ -64,20 +72,24 @@ internal class TransactionBodyAdapter : RecyclerView.Adapter<TransactionPayloadV
         }
     }
 
-    internal fun highlightQueryWithColors(newText: String, backgroundColor: Int, foregroundColor: Int) {
+    internal fun highlightQueryWithColors(
+        newText: String,
+        backgroundColor: Int,
+        foregroundColor: Int
+    ) {
```
Comment (ArjanSM) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadAdapter.kt:79]: nit: revert this change

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt [hunk 6]
```diff
@@ -250,7 +254,7 @@ internal class TransactionPayloadFragment :
             if (type == PayloadType.RESPONSE && responseBitmap != null) {
                 val bitmapLuminance = responseBitmap.calculateLuminance()
                 result.add(TransactionPayloadItem.ImageItem(responseBitmap, bitmapLuminance))
-                return@withContext result
+                return@withContext  result
```
Comment (ArjanSM) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt:257]: nit: remove the extra whitespace
```suggestion
return@withContext result

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt [hunk 7]
```diff
@@ -263,15 +267,20 @@ internal class TransactionPayloadFragment :
...
             return@withContext result
         }
     }
 
-    private suspend fun saveToFile(type: PayloadType, uri: Uri, transaction: HttpTransaction): Boolean {
+    private suspend fun saveToFile(
+        type: PayloadType,
+        uri: Uri,
+        transaction: HttpTransaction
+    ): Boolean {
```
Comment (ArjanSM) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt:283]: nit: may be revert the formatting.

**Review by ArjanSM — COMMENTED** [review]
sorry for the delay. Just a few to nits to begin with. I'll give it a more thorough pass later in the week.

**Comment by Amirhy** [comment]
> sorry for the delay. Just a few to nits to begin with. I'll give it a more thorough pass later in the week.

**Comment by Amirhy** [comment]
> sorry for the delay. Just a few to nits to begin with. I'll give it a more thorough pass later in the week.

Hi @ArjanSM, 

Hope you are doing well.

I resolved all threads you created.Please let me know if you have any other comment.

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/support/SpanTextUtil.kt [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+            }
+            val sb = SpannableStringBuilder()
+            printifyRecursive(jsonElement, StringBuilder(""), sb)
+            return sb
+        }
+
+        private fun printifyRecursive(
+            transformedJson: JsonElement,
+            currentIndent: StringBuilder,
+            sb: SpannableStringBuilder
```
Comment (ArjanSM) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/support/SpanTextUtil.kt:35]: nit: we could maintain a consistent order of arguments across all printfy* methods
``` suggestion
        private fun printifyRecursive(
            sb: SpannableStringBuilder,
            currentIndent: StringBuilder,
            transformedJson: JsonElement
            )

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/support/SpanTextUtil.kt [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+            transformedJson: JsonElement
+        ) {
+            if (transformedJson.asJsonArray.isEmpty) {
+                sb.appendWithColor(
+                    "[]",
+                    JSON_SIGN_ELEMENTS_COLOR
+                )
+                return
+            }
+            sb.appendWithColor("$indent[\n", JSON_SIGN_ELEMENTS_COLOR)
```
Comment (ArjanSM) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/support/SpanTextUtil.kt:58]: nit: for the sake of consistency
```suggestion
            sb.appendWithColor("${indent}[\n", JSON_SIGN_ELEMENTS_COLOR)

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/support/SpanTextUtil.kt [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+                    printifyRecursive(item, indent, sb)
+                else {
+                    sb.append(indent)
+                    sb.appendJsonValue(item)
+                }
+                if (index != transformedJson.asJsonArray.size() - 1)
+                    sb.appendWithColor(",", JSON_SIGN_ELEMENTS_COLOR).append("\n")
+            }
+            val finalIndent = StringBuilder(indent.dropLast(2))
+            sb.appendWithColor("\n$finalIndent]", JSON_SIGN_ELEMENTS_COLOR)
```
Comment (ArjanSM) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/support/SpanTextUtil.kt:72]: nit: wrap `${finalIndent}`

**Review by ArjanSM — COMMENTED** [review]
Getting there! 
Just a bunch of few nits.

File: library/src/androidTest/kotlin/com/chuckerteam/chucker/SpanUtilTest.kt [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+import org.junit.Assert
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+public class SpanUtilTest {
+    @SuppressLint("CheckResult")
+    @Test
+    public fun json_can_have_null_value() {
+
```
[the remark this answers is not in the document]
Comment (Amirhy) [reply library/src/androidTest/kotlin/com/chuckerteam/chucker/SpanUtilTest.kt]: Hi @ArjanSM 

Hope you are doing well. I want to inform you that I changed the method's name in my test class and again I am using underscore instead of space as this convention is not supported in android runtime tests.
You can check [this link](https://kotlinlang.org/docs/coding-conventions.html#names-for-test-methods) for more information.

**Comment by Amirhy** [comment]
Hi Dear @ArjanSM 

Do you have any new nits about my changes?  ;-) 

I want you check [this comment](https://github.com/ChuckerTeam/chucker/pull/930#discussion_r1106963903) and let me know what is your opinion?

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/support/SpanTextUtil.kt [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+import com.google.gson.JsonParser
+import com.google.gson.JsonSyntaxException
+
+
+public class SpanTextUtil {
+    public companion object {
+        private val JSON_KEY_COLOR = Color.parseColor("#8B0057")
+        private val JSON_STRING_VALUE_COLOR = Color.parseColor("#2F00FF")
+        private val JSON_DIGIT_AND_NULL_VALUE_COLOR = Color.parseColor("#E84B31")
+        private val JSON_SIGN_ELEMENTS_COLOR = Color.parseColor("#474747")
```
Comment (cortinico) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/support/SpanTextUtil.kt:18]: Can we get those not hardcoded but moved to resource + have a `-night` variants with brighter and more pale colors?

That's how it looks like on dark mode:
Screenshot_1677346348

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/support/SpanTextUtil.kt [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+        public fun spanJson(input: CharSequence): SpannableStringBuilder {
+            val jsonElement = try {
+                JsonParser.parseString(input.toString())
+            } catch (e: JsonSyntaxException) {
+                Logger.warn("Json structure is invalid so it can not be formatted", e)
+                return SpannableStringBuilder.valueOf(input)
+            }
+            val sb = SpannableStringBuilder()
+            printifyRecursive(sb, StringBuilder(""), jsonElement)
+            return sb
```
Comment (cortinico) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/support/SpanTextUtil.kt:29]:
```suggestion
            return SpannableStringBuilder().also {
                printifyRecursive(it, StringBuilder(""), jsonElement)  
            }
```

That's a nit

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt [hunk 14]
```diff
@@ -311,4 +316,17 @@ internal class TransactionPayloadFragment :
...
+    private fun CharSequence.lines(): List<CharSequence> {
+        val linesList = this.lineSequence().toList()
+        val result = mutableListOf<CharSequence>()
+        var lineIndex = 0
+        for (index in linesList.indices) {
+            result.add(subSequence(lineIndex, lineIndex + linesList[index].length))
+            lineIndex += linesList[index].length + 1
+        }
+        if (result.isEmpty())
+            result.add(subSequence(0, length))
```
Comment (cortinico) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt:329]: Why this?

**Review by cortinico — COMMENTED** [review]
Code looks good to me 👍 @ArjanSM already did a great review! 
Only left a minor nits.

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt (hunk 14 above)
Comment (Amirhy) [reply library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt:329]: That's because when the char sequence doesn't have any lines so I should add all of its content to the result.

**Comment by ArjanSM** [comment]
> Hi Dear @ArjanSM
> 
> Do you have any new nits about my changes? ;-)
> 
> I want you check [this comment](https://github.com/ChuckerTeam/chucker/pull/930#discussion_r1106963903) and let me know what is your opinion?

Aah! my bad! 👍

**Review by ArjanSM — APPROVED** [review]

**Review by cortinico — APPROVED** [review]
This is awesome @Amirhy Thanks for taking the time to add this feature!

**Comment by misael-jonathan** [comment]
hi @Amirhy , I think there's performance issue when we used this implementation in large json data. I just updated my chucker with this colored request/response, in some cases it took 2-3 secs to load my response, and if its super big it'll just stuck on loading state. Can u verify this issue on your side?

**Comment by Amirhy** [comment]
Hi @misael-jonathan , 
Thanks for sharing your experience.
If it is possible please send me your json to my email (amirhosein.hosseini76@gmail.com)
I will investigate on it and inform you as soon as possible.

**Comment by misael-jonathan** [comment]
Hi, I just sent the json to your email. Actually it is the same json you can find in LargeJson.kt in the sample app. 
So, I think you can start from there.

Thanks @Amirhy
